WIP Import user defined objects as modules

NDagestad opened the pull request:
This PR is on my packaging branch/PR.

This uses importlib's package metadata to make fuddly automagically discover data_models/projects and other user defined objects.

When a user defines a data_model{,_strategy} or project, they can add a [project.entry-points."<group-name>"] section in their pyporject.toml file. Fuddly can then request from importlib, all entry_points of a specific group-name.
2 group name were chose for fuddly:
- fuddly.data_models
- fuddly.projects

because of this, the requirement for project files to be named with a '_proj' suffix has been lifted.
If a project does have the suffix, it will be removed in the project name in fuddly's internal workings though.

The internal data_models/projects have been exposed in the fuddly's pyproject.toml file, and are therefore imported as modules instead of being looked up directly in the filesystem.

TODO

- [x] Documents the new behaviour
- [x] Merge data_models and data_models_strategies
        For now, a data_model, and data_model's strategy must have the same entrypoint name so fuddly can associate them. A better way of doing this would be if the strategies were added to an array of the data_model (similarly to how targets are defined for projects)
- [x] Check the filsystem before modules and warn but ignore duplicates
        This way, way can override a module by creating a local object with the same name.
- [x] Make fuddly detect if it is being run from a module or directly and import the internal object from the filesystem if ran directly
- [x] Put the entry point group names in a configuration or at least in a globally accessible variable

NDagestad committed:
update loading logic

Also revert most of the style changes from https://github.com/6E006B/fuddly

NDagestad committed:
Make user_{data_models,targets,info,projects} loadable from a python module

Using importlibs entry_point package metadata functionality, any package
that declares entry points in the fuddly.data_models and fuddly.projects
groups will have them be available to fuddly when installed.

NDagestad committed:
Fix the names of internal projects in the pyproject metadata

NDagestad committed:
Make pyporject use requirements.txt (generated by pipenv)

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -1071,6 +1274,7 @@ def _add_project(self, project, targets, logger, prj_rld_args, reload_prj=False)
             self.prj_list.remove(prj)
             self.prj_list.append(project)
         else:
+            # TODO: Catch this hieght up and ignore double imports ?
```
higher up ?

PoneyUHC reviewed (changes requested):
typo

NDagestad committed:
Move the entry_point group names to a global variable

NDagestad committed:
Ignore duplicate imports of project and data models from modules

This is so that it is possible to override modules by filesystem
objects

NDagestad committed:
Make {data_model,project}.name the same as module.name

NDagestad committed:
Remove unneeded imports

NDagestad committed:
Exit only if launched as a script

NDagestad committed:
Remove old import magic

NDagestad committed:
Move fuddly_shell back in the fuddly package

commit 272efe4bc8708aa3 got revertd somehow...

NDagestad committed:
Import internal objects from the fs only if ran from the sources directly

NDagestad committed:
Start documenting the packaging steps

NDagestad committed:
Merge _import_dm_from functions and make data_models+strategies be a module

NDagestad committed:
Reorganize data_models

Each data model is now a module:
    <name>.py -> <name>/__init__.py
    <name>_strategy.py -> <name>/strategy.py

NDagestad committed:
Update fs import logic for new layout

NDagestad committed:
Bunch of missing files because of a surprising .gitignore

NDagestad committed:
Fix the import logic for projects from the filesystem with the _proj suffix

NDagestad committed:
Rename project files to remove _proj suffix

NDagestad committed:
Update pyproject entrypoints after renamint of projects

NDagestad committed:
Update requirements.txt from Pipfile (manually)

NDagestad committed:
Try to workaround importlib weirdness

If fuddly.egg-info folder is in the src/ folder.
importlib.metadata.files will return and empty array ?

NDagestad committed:
Error message should have been an f-string

NDagestad committed:
Update documentation

NDagestad committed:
Update documentation

NDagestad committed:
Useless line + removing old unused entry_point group name

NDagestad committed:
Use == instead of in to do a normal compare

The previous code would look for the string being contained in the other
one, leading to possible false positives

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -59,19 +59,18 @@
 sys.path.insert(0, fuddly_data_folder)
 sys.path.insert(0, external_libs_folder)
 
+#TODO These are not needed right ?
```
to resolve

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -759,125 +758,159 @@ def monitor_probes(self, prefix=None, force_record=False):
+                for m in modules:
+                    relpath = dirname(m)[len(base_path)+1:]
+                    key=prefix+relpath
+                    if data_models.get(key) is None:
```
use defaultdict ?

NDagestad replied:
For what, data_model ? It is a `colloections.OrderedDict()` (based on what was used before)

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -759,125 +758,159 @@ def monitor_probes(self, prefix=None, force_record=False):
+
         data_models = collections.OrderedDict()
 
         def populate_data_models(path, prefix=""):
```
why defining a function inside another to use it only once
--> get out or remove function

NDagestad committed:
Add a better local source execution detection

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -759,125 +758,159 @@ def monitor_probes(self, prefix=None, force_record=False):
+
         data_models = collections.OrderedDict()
 
         def populate_data_models(path, prefix=""):
```
not big fan of function in another, adds useless indentation (5-6 levels here)

NDagestad replied:
Not my addition, it was already like this before so I didn't change it

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -759,125 +758,159 @@ def monitor_probes(self, prefix=None, force_record=False):
+                    if fmkDB_update:
+                        # populate FMK DB
+                        self._fmkDB_insert_dm_and_dmakers(dm_params["dm"].name, dm_params["tactics"])
+                else:
```
if dm_params is None: false break

NDagestad replied:
fmkDB_update is a boolean value?
dm_params is checked to not be None before entering the block?
I don't get what you are saying

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -759,125 +758,159 @@ def monitor_probes(self, prefix=None, force_record=False):
+                if fmkDB_update:
+                    # populate FMK DB
+                    self._fmkDB_insert_dm_and_dmakers(dm_params['dm'].name, dm_params['tactics'])
+            else:
```
same as before, if None: =false; break

NDagestad replied:
Same as before, fmkDB_update is a boolean

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -963,21 +1004,60 @@ def populate_projects(path, prefix=""):
+                                  reload_prj=False)
+                if fmkDB_update:
+                    self.fmkDB.insert_project(prj_params["project"].name)
+            else:
```
same

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -963,21 +1004,60 @@ def populate_projects(path, prefix=""):
+                self.print(colorize(f"*** Problem during reload of '{name}.py'/'{name}_proj.py' ***", rgb=Color.ERROR))
             else:
-                self.print(colorize("*** Problem during import of '%s_proj.py' ***" % (name), rgb=Color.ERROR))
+                self.print(colorize(f"*** Problem during import of '{name}.py'/'{name}_proj.py' ***", rgb=Color.ERROR))
```
both branches are the same ??

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -1065,7 +1233,7 @@ def _add_project(self, project, targets, logger, prj_rld_args, reload_prj=False)
                     break
             else:
-                raise ValueError
+                raise ValueError(porject.name)
```
c'est  toi le porject

NDagestad replied:
🐷

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -1056,6 +1136,94 @@ def _import_project(self, prefix, name, reload_prj=False):
+            logger.name = name
+
+        # Targets
+        target=[]
```
wrong spelling, rip if error

NDagestad replied:
Python is magic, so it still works 🪄

PoneyUHC commented on src/fuddly/framework/plumbing.py:
```diff
@@ -1056,6 +1136,94 @@ def _import_project(self, prefix, name, reload_prj=False):
+
+        return {
+            "project": project,
+            "target": targets,
```
"target" or "targets" ?

NDagestad replied:
"target" = targets is correct

NDagestad committed:
Fix comments from PR

NDagestad committed:
Doc tweak

NDagestad committed:
Change the data_model structure

The new structure is:

module_name
├── dm.py
├── __init__.py
└── strategy.py

where __init__.py imports and exposes data_model and tactics from the
rest of the module

NDagestad committed:
Small syntax changes and log message change

NDagestad committed:
Update documentation
